# adapted from page bottom of:
# http://www.molgen.ua.ac.be/bioinfo/acourse/Bioconductor-by-examples.txt

# see also:
#http://bioconductor.fhcrc.org/packages/release/bioc/vignettes/Biostrings/inst/doc/MultipleAlignments.R

# Some web services based packages
# ================================

#some useful packages
install.packages("RCurl")
library(RCurl)
source("https://bioconductor.org/biocLite.R")
biocLite()
biocLite("Biostrings")
library(Biostrings)
install.packages("rentrez")
library(rentrez)

setwd("~/Desktop")

#Here is the FASTA sequence provided in the BLAST assignment:
seq <- "HLYPGEVCPGMDIRNNLTRLHELENCSVIEGHLQILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYALVIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRILDSVEDNHIVLNKDDNEECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECLGNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQGCHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGCTVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRSYALVSLSFFRKLRLIRGETLEIGNYSFYALDNQNLRQLWDWSKHNLTITQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQERNDIALKTNGDKASCENELLKFSYIRTSFDKIS"

#After finding the best hit that is actual, not predicted sequence, we can use its accession number
seq1 <- "NP_000199.2"
seq2 <- "NP_963860.1"

# blast a sequence using the NCBI REST interface
# enable low-complexity filtering with FILTER="L", whereas FILTER="F" turns filltering off
# FORMAT_TYPE="Text" vs "HTML"
# QUERY =  vs using source from internet like "http://www.ncbi.nlm.nih.gov/blast/Blast.cgi?QUERY=555
# use NCBI GI numbers in the output page with NCBI_GI="on"  
ncbi_blast = function(seq,database="nr",expect=10,program="blastp",hitlistsize=100, filter="F", matrix="BLOSUM90", word_size=6, format_type="HTML") {
  # based on the NCBI REST interface (https://ncbi.github.io/blast-cloud/dev/api.html)
  job = as.character(getForm("https://blast.ncbi.nlm.nih.gov/Blast.cgi", QUERY=seq, DATABASE=database, MATRIX=matrix,
                             HITLIST_SIZE=hitlistsize, FILTER=filter, EXPECT=expect, FORMAT_TYPE=format_type, PROGRAM=program, CLIENT="web",
                             WORD_SIZE=word_size, SERVICE="plain", NCBI_GI="on", CMD="Put"))
  m = regexpr("RID = ([^\n]+)",job)
  id = substring(job,m[1]+6,m[1]+attributes(m)$match.length-1)
  m = regexpr("RTOE = ([^\n]+)",job)
  rtoe = substring(job,m[1]+7,m[1]+attributes(m)$match.length-1)
  result = "Status=WAITING"
  while (length(grep("Status=WAITING",result))) {
    #Sys.sleep(1)
    Sys.sleep(60)
    result = as.character(getForm("https://blast.ncbi.nlm.nih.gov/Blast.cgi", RID=id, FORMAT_TYPE="Text", CMD="Get"))
  }
  getForm("https://blast.ncbi.nlm.nih.gov/Blast.cgi", RID=id, CMD="Delete")
  return(result)
}
b1 = ncbi_blast(seq1,database="refseq_protein",program="blastp", matrix="BLOSUM45") 
#b = ncbi_blast(seq,database="nr",program="PSI-BLAST") 
b2 = ncbi_blast(seq1,database="refseq_rna",program="tblastn")
b3 = ncbi_blast(seq2,database="refseq_protein",program="blastp",filter="L")

#write.table(b, "blast_data.html", sep="\t")
write.table(b1, "blast_data#1_distantHomologs.html", sep="\t")
write.table(b2, "blast_data#2_nucleotideSeq.html", sep="\t")
write.table(b3, "blast_data#3_SLC16A13Homologs.html", sep="\t")

# #alt save output to file
#sink("blast_data.html")
sink("blast_data#1_distantHomologs.html")
sink("blast_data#2_nucleotideSeq.html")
sink("blast_data#3_SLC16A13Homologs.html")

cat(b1)
cat(b2)
cat(b3)

sink()
# 
# #alt save output to file
#fileConn<-file("output.txt")
#fileConn<-file("blast_data#1_distantHomologs.txt")
#fileConn<-file("blast_data#2_nucleotideSeq.txt")
#fileConn<-file("blast_data#3_SLC16A13Homologs.txt")

#writeLines(b1, fileConn)
#writeLines(b2, fileConn)
#writeLines(b3, fileConn)

#close(fileConn)

#ToDo:
#1
#Starting with the accession number of the protein that was your best hit of actual
# (not predicted) sequences in the above search
# perform a new search for distantly related homologs

#2
# Find the accession number of the nucleotide sequence that corresponds with the
# protein sequence that you used as query above. Retrieve the pairwise alignments
# ot the best candidates as aligned with your query

#NM_000208.3
#ALIGNMENTS
#>NM_000208.3 Homo sapiens insulin receptor (INSR), transcript variant 1, mRNA
#Length=9375

# Score = 2884 bits (7476),  Expect = 0.0, Method: Compositional matrix adjust.
# Identities = 1382/1382 (100%), Positives = 1382/1382 (100%), Gaps = 0/1382 (0%)
# Frame = +1
# 
# Query  1     MATGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLTRLHELENCSVIEGHL  60
# MATGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLTRLHELENCSVIEGHL
# Sbjct  412   MATGGRRGAAAAPLLVAVAALLLGAAGHLYPGEVCPGMDIRNNLTRLHELENCSVIEGHL  591
# 
# Query  61    QILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYAL  120
# QILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYAL
# Sbjct  592   QILLMFKTRPEDFRDLSFPKLIMITDYLLLFRVYGLESLKDLFPNLTVIRGSRLFFNYAL  771
# 
# Query  121   VIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRILDSVEDNYIVLNKDDNE  180
# VIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRILDSVEDNYIVLNKDDNE
# Sbjct  772   VIFEMVHLKELGLYNLMNITRGSVRIEKNNELCYLATIDWSRILDSVEDNYIVLNKDDNE  951
# 
# Query  181   ECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECL  240
# ECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECL
# Sbjct  952   ECGDICPGTAKGKTNCPATVINGQFVERCWTHSHCQKVCPTICKSHGCTAEGLCCHSECL  1131
# 
# Query  241   GNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQG  300
# GNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQG
# Sbjct  1132  GNCSQPDDPTKCVACRNFYLDGRCVETCPPPYYHFQDWRCVNFSFCQDLHHKCKNSRRQG  1311
# 
# Query  301   CHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGC  360
# CHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGC
# Sbjct  1312  CHQYVIHNNKCIPECPSGYTMNSSNLLCTPCLGPCPKVCHLLEGEKTIDSVTSAQELRGC  1491
# 
# Query  361   TVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRSYALVSLSFFRKLRLIRGETL  420
# TVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRSYALVSLSFFRKLRLIRGETL
# Sbjct  1492  TVINGSLIINIRGGNNLAAELEANLGLIEEISGYLKIRRSYALVSLSFFRKLRLIRGETL  1671
# 
# Query  421   EIGNYSFYALDNQNLRQLWDWSKHNLTITQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQE  480
# EIGNYSFYALDNQNLRQLWDWSKHNLTITQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQE
# Sbjct  1672  EIGNYSFYALDNQNLRQLWDWSKHNLTITQGKLFFHYNPKLCLSEIHKMEEVSGTKGRQE  1851
# 
# Query  481   RNDIALKTNGDQASCENELLKFSYIRTSFDKILLRWEPYWPPDFRDLLGFMLFYKEAPYQ  540
# RNDIALKTNGDQASCENELLKFSYIRTSFDKILLRWEPYWPPDFRDLLGFMLFYKEAPYQ
# Sbjct  1852  RNDIALKTNGDQASCENELLKFSYIRTSFDKILLRWEPYWPPDFRDLLGFMLFYKEAPYQ  2031
# 
# Query  541   NVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHPGWLMRGLKPWTQYAIFVKTLVTFS  600
# NVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHPGWLMRGLKPWTQYAIFVKTLVTFS
# Sbjct  2032  NVTEFDGQDACGSNSWTVVDIDPPLRSNDPKSQNHPGWLMRGLKPWTQYAIFVKTLVTFS  2211
# 
# Query  601   DERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSSSQIILKWKPPSDPNGNITHYLVFWE  660
# DERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSSSQIILKWKPPSDPNGNITHYLVFWE
# Sbjct  2212  DERRTYGAKSDIIYVQTDATNPSVPLDPISVSNSSSQIILKWKPPSDPNGNITHYLVFWE  2391
# 
# Query  661   RQAEDSELFELDYCLKGLKLPSRTWSPPFESEDSQKHNQSEYEDSAGECCSCPKTDSQIL  720
# RQAEDSELFELDYCLKGLKLPSRTWSPPFESEDSQKHNQSEYEDSAGECCSCPKTDSQIL
# Sbjct  2392  RQAEDSELFELDYCLKGLKLPSRTWSPPFESEDSQKHNQSEYEDSAGECCSCPKTDSQIL  2571
# 
# Query  721   KELEESSFRKTFEDYLHNVVFVPRKTSSGTGAEDPRPSRKRRSLGDVGNVTVAVPTVAAF  780
# KELEESSFRKTFEDYLHNVVFVPRKTSSGTGAEDPRPSRKRRSLGDVGNVTVAVPTVAAF
# Sbjct  2572  KELEESSFRKTFEDYLHNVVFVPRKTSSGTGAEDPRPSRKRRSLGDVGNVTVAVPTVAAF  2751
# 
# Query  781   PNTSSTSVPTSPEEHRPFEKVVNKESLVISGLRHFTGYRIELQACNQDTPEERCSVAAYV  840
# PNTSSTSVPTSPEEHRPFEKVVNKESLVISGLRHFTGYRIELQACNQDTPEERCSVAAYV
# Sbjct  2752  PNTSSTSVPTSPEEHRPFEKVVNKESLVISGLRHFTGYRIELQACNQDTPEERCSVAAYV  2931
# 
# Query  841   SARTMPEAKADDIVGPVTHEIFENNVVHLMWQEPKEPNGLIVLYEVSYRRYGDEELHLCV  900
# SARTMPEAKADDIVGPVTHEIFENNVVHLMWQEPKEPNGLIVLYEVSYRRYGDEELHLCV
# Sbjct  2932  SARTMPEAKADDIVGPVTHEIFENNVVHLMWQEPKEPNGLIVLYEVSYRRYGDEELHLCV  3111
# 
# Query  901   SRKHFALERGCRLRGLSPGNYSVRIRATSLAGNGSWTEPTYFYVTDYLDVPSNIAKIIIG  960
# SRKHFALERGCRLRGLSPGNYSVRIRATSLAGNGSWTEPTYFYVTDYLDVPSNIAKIIIG
# Sbjct  3112  SRKHFALERGCRLRGLSPGNYSVRIRATSLAGNGSWTEPTYFYVTDYLDVPSNIAKIIIG  3291
# 
# Query  961   PLIFVFLFSVVIGSIYLFLRKRQPDGPLGPLYASSNPEYLSASDVFPCSVYVPDEWEVSR  1020
# PLIFVFLFSVVIGSIYLFLRKRQPDGPLGPLYASSNPEYLSASDVFPCSVYVPDEWEVSR
# Sbjct  3292  PLIFVFLFSVVIGSIYLFLRKRQPDGPLGPLYASSNPEYLSASDVFPCSVYVPDEWEVSR  3471
# 
# Query  1021  EKITLLRELGQGSFGMVYEGNARDIIKGEAETRVAVKTVNESASLRERIEFLNEASVMKG  1080
# EKITLLRELGQGSFGMVYEGNARDIIKGEAETRVAVKTVNESASLRERIEFLNEASVMKG
# Sbjct  3472  EKITLLRELGQGSFGMVYEGNARDIIKGEAETRVAVKTVNESASLRERIEFLNEASVMKG  3651
# 
# Query  1081  FTCHHVVRLLGVVSKGQPTLVVMELMAHGDLKSYLRSLRPEAENNPGRPPPTLQEMIQMA  1140
# FTCHHVVRLLGVVSKGQPTLVVMELMAHGDLKSYLRSLRPEAENNPGRPPPTLQEMIQMA
# Sbjct  3652  FTCHHVVRLLGVVSKGQPTLVVMELMAHGDLKSYLRSLRPEAENNPGRPPPTLQEMIQMA  3831
# 
# Query  1141  AEIADGMAYLNAKKFVHRDLAARNCMVAHDFTVKIGDFGMTRDIYETDYYRKGGKGLLPV  1200
# AEIADGMAYLNAKKFVHRDLAARNCMVAHDFTVKIGDFGMTRDIYETDYYRKGGKGLLPV
# Sbjct  3832  AEIADGMAYLNAKKFVHRDLAARNCMVAHDFTVKIGDFGMTRDIYETDYYRKGGKGLLPV  4011
# 
# Query  1201  RWMAPESLKDGVFTTSSDMWSFGVVLWEITSLAEQPYQGLSNEQVLKFVMDGGYLDQPDN  1260
# RWMAPESLKDGVFTTSSDMWSFGVVLWEITSLAEQPYQGLSNEQVLKFVMDGGYLDQPDN
# Sbjct  4012  RWMAPESLKDGVFTTSSDMWSFGVVLWEITSLAEQPYQGLSNEQVLKFVMDGGYLDQPDN  4191
# 
# Query  1261  CPERVTDLMRMCWQFNPKMRPTFLEIVNLLKDDLHPSFPEVSFFHSEENKAPESEELEME  1320
# CPERVTDLMRMCWQFNPKMRPTFLEIVNLLKDDLHPSFPEVSFFHSEENKAPESEELEME
# Sbjct  4192  CPERVTDLMRMCWQFNPKMRPTFLEIVNLLKDDLHPSFPEVSFFHSEENKAPESEELEME  4371
# 
# Query  1321  FEDMENVPLDRSSHCQREEAGGRDGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSN  1380
# FEDMENVPLDRSSHCQREEAGGRDGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSN
# Sbjct  4372  FEDMENVPLDRSSHCQREEAGGRDGGSSLGFKRSYEEHIPYTHMNGGKKNGRILTLPRSN  4551
# 
# Query  1381  PS  1382
# PS
# Sbjct  4552  PS  4557


#3
#Find homologs to the SLC16A13 protein (paralog of the query protein) 
# that you used in part 1
